File: ios/FlutLink/UI/Admin/AdminViewModel.swift
// TEST — iOS port created by opencode. Not a production build.
import Foundation

@MainActor
final class AdminViewModel: ObservableObject {
    private let sessionManager: SessionManager

    @Published var users: [ManagedUser] = []
    @Published var loading = false
    @Published var errorMessage: String?
    @Published var search = ""
    @Published var hasMore = false

    private var offset = 0
    private var searchTerm = ""

    var session: AuthSession? { sessionManager.session }

    init(sessionManager: SessionManager) {
        self.sessionManager = sessionManager
    }

    func refresh() { loadUsers() }

    func loadUsers() {
        guard let s = session else { return }
        loading = true; errorMessage = nil
        users = []; hasMore = false
        searchTerm = search.trimmingCharacters(in: .whitespaces)
        offset = 0
        Task {
            do {
                try await loadPage(s, append: false)
            } catch {
                errorMessage = LoginViewModel.mapError(error)
            }
            loading = false
        }
    }

    func loadMore() {
        guard let s = session, !loading else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                try await loadPage(s, append: true)
            } catch {
                errorMessage = LoginViewModel.mapError(error)
            }
            loading = false
        }
    }

    private func loadPage(_ s: AuthSession, append: Bool) async throws {
        let api = FlutCloudApi(session: s)
        let page = try await api.listUsersPage(search: searchTerm, offset: offset)
        var managed: [ManagedUser] = []
        for id in page {
            if let user = try? await api.getUser(userId: id) {
                managed.append(user)
            } else {
                managed.append(ManagedUser(id: id, displayName: nil, email: nil, quota: nil, groups: [], enabled: true))
            }
        }
        users = append ? users + managed : managed
        hasMore = page.count == 200
        offset += page.count
    }

    func createUser(userId: String, password: String, displayName: String?) {
        guard !userId.trimmingCharacters(in: .whitespaces).isEmpty, !password.isEmpty, let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                try await FlutCloudApi(session: s).createUser(userId: userId.trimmingCharacters(in: .whitespaces), password: password, displayName: displayName?.trimmingCharacters(in: .whitespaces).isEmpty == true ? nil : displayName)
                loadUsers()
            } catch {
                errorMessage = LoginViewModel.mapError(error)
            }
            loading = false
        }
    }

    func deleteUser(_ user: ManagedUser) {
        guard let s = session else { return }
        if user.id == s.username {
            errorMessage = NSLocalizedString("cannot_delete_self", comment: "")
            return
        }
        loading = true; errorMessage = nil
        Task {
            do {
                try await FlutCloudApi(session: s).deleteUser(userId: user.id)
                users.removeAll { $0.id == user.id }
            } catch {
                errorMessage = LoginViewModel.mapError(error)
            }
            loading = false
        }
    }

    func setQuota(_ user: ManagedUser, quotaBytes: Int64?) {
        guard let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                try await FlutCloudApi(session: s).setUserQuota(userId: user.id, quotaBytes: quotaBytes)
                loadUsers()
            } catch {
                errorMessage = LoginViewModel.mapError(error)
            }
            loading = false
        }
    }

    func setEnabled(_ user: ManagedUser, enabled: Bool) {
        guard let s = session else { return }
        if user.id == s.username && !enabled {
            errorMessage = NSLocalizedString("cannot_disable_self", comment: "")
            return
        }
        loading = true; errorMessage = nil
        Task {
            do {
                try await FlutCloudApi(session: s).updateUser(userId: user.id, key: "enabled", value: enabled ? "1" : "0")
                loadUsers()
            } catch {
                errorMessage = LoginViewModel.mapError(error)
            }
            loading = false
        }
    }

    func addToGroup(_ user: ManagedUser, group: String) {
        let g = group.trimmingCharacters(in: .whitespaces)
        guard !g.isEmpty, let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                try await FlutCloudApi(session: s).addGroupMember(groupId: g, userId: user.id)
                loadUsers()
            } catch {
                errorMessage = LoginViewModel.mapError(error)
            }
            loading = false
        }
    }

    func removeFromGroup(_ user: ManagedUser, group: String) {
        guard let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                try await FlutCloudApi(session: s).removeGroupMember(groupId: group, userId: user.id)
                loadUsers()
            } catch {
                errorMessage = LoginViewModel.mapError(error)
            }
            loading = false
        }
    }

    func createGroup(_ name: String) {
        let g = name.trimmingCharacters(in: .whitespaces)
        guard !g.isEmpty, let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                try await FlutCloudApi(session: s).createGroup(groupId: g)
            } catch {
                errorMessage = LoginViewModel.mapError(error)
            }
            loading = false
        }
    }

    func clearError() { errorMessage = nil }
}

File: kmp/shared/src/androidMain/kotlin/com/flutcloud/flutlink/ui/viewmodel/AdminViewModel.kt
package com.flutcloud.flutlink.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.flutcloud.flutlink.AppContainer
import com.flutcloud.flutlink.data.ApiException
import com.flutcloud.flutlink.data.AuthSession
import com.flutcloud.flutlink.data.NetworkException
import com.flutcloud.flutlink.data.dto.ManagedUser
import com.flutcloud.flutlink.ui.UiMessage
import com.flutcloud.flutlink.ui.networkUiMessage
import com.flutcloud.flutlink.ui.toUiMessage
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class AdminViewModel(private val container: AppContainer) : ViewModel() {

    private val session get() = container.sessionManager.session.value

    val users = MutableStateFlow<List<ManagedUser>>(emptyList())
    val loading = MutableStateFlow(false)
    val error = MutableStateFlow<UiMessage?>(null)
    val search = MutableStateFlow("")
    val hasMore = MutableStateFlow(false)

    private var offset = 0
    private var searchTerm = ""

    fun refresh() = loadUsers()

    /** Reset the list when the search field is empty (no server round-trips). */
    fun clearSearch() {
        users.value = emptyList()
        hasMore.value = false
        searchTerm = ""
        offset = 0
        loading.value = false
        error.value = null
    }

    fun loadUsers() {
        val s = session ?: return
        val term = search.value.trim()
        if (term.isEmpty()) {
            searchTerm = ""
            offset = 0
            users.value = emptyList()
            hasMore.value = false
            loading.value = false
            return
        }
        searchTerm = term
        offset = 0
        viewModelScope.launch {
            loading.value = true
            error.value = null
            users.value = emptyList()
            hasMore.value = false
            try {
                loadPage(s, append = false)
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            } finally {
                loading.value = false
            }
        }
    }

    fun loadMore() {
        val s = session ?: return
        if (loading.value) return
        viewModelScope.launch {
            loading.value = true
            error.value = null
            try {
                loadPage(s, append = true)
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            } finally {
                loading.value = false
            }
        }
    }

    private suspend fun loadPage(s: AuthSession, append: Boolean) {
        val page = container.ocsApi.listUsersPage(s, searchTerm, offset)
        val managed = page.map { id ->
            runCatching { container.ocsApi.getUser(s, id) }
                .getOrElse { ManagedUser(id = id, displayName = null, email = null, quota = null, groups = emptyList(), enabled = true) }
        }
        users.value = if (append) users.value + managed else managed
        hasMore.value = page.size == 200
        offset += page.size
    }

    fun createUser(userId: String, password: String, displayName: String?) {
        if (userId.isBlank() || password.isBlank()) return
        val s = session ?: return
        viewModelScope.launch {
            error.value = null
            try {
                container.ocsApi.createUser(s, userId.trim(), password, displayName?.trim()?.ifBlank { null })
                loadUsers()
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    fun deleteUser(user: ManagedUser) {
        val s = session ?: return
        if (user.id == s.username) {
            error.value = UiMessage(com.flutcloud.flutlink.R.string.cannot_delete_self)
            return
        }
        viewModelScope.launch {
            error.value = null
            try {
                container.ocsApi.deleteUser(s, user.id)
                users.value = users.value.filterNot { it.id == user.id }
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    fun setQuota(user: ManagedUser, quotaBytes: Long?) {
        val s = session ?: return
        viewModelScope.launch {
            error.value = null
            try {
                container.ocsApi.setUserQuota(s, user.id, quotaBytes)
                loadUsers()
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    fun setEnabled(user: ManagedUser, enabled: Boolean) {
        val s = session ?: return
        if (user.id == s.username && !enabled) {
            error.value = UiMessage(com.flutcloud.flutlink.R.string.cannot_disable_self)
            return
        }
        viewModelScope.launch {
            error.value = null
            try {
                container.ocsApi.updateUser(s, user.id, "enabled", if (enabled) "1" else "0")
                loadUsers()
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    /** Add `user` to the group `group` (creates the membership on the server). */
    fun addToGroup(user: ManagedUser, group: String) {
        val g = group.trim()
        if (g.isEmpty()) return
        val s = session ?: return
        viewModelScope.launch {
            error.value = null
            try {
                container.ocsApi.addGroupMember(s, g, user.id)
                loadUsers()
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    /** Remove `user` from the group `group`. */
    fun removeFromGroup(user: ManagedUser, group: String) {
        val s = session ?: return
        viewModelScope.launch {
            error.value = null
            try {
                container.ocsApi.removeGroupMember(s, group, user.id)
                loadUsers()
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    /** Create a new group with the given id. */
    fun createGroup(name: String) {
        val g = name.trim()
        if (g.isEmpty()) return
        val s = session ?: return
        viewModelScope.launch {
            error.value = null
            try {
                container.ocsApi.createGroup(s, g)
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    fun clearError() {
        error.value = null
    }
}